### PR TITLE
Add in rprop solver

### PR DESCRIPTION
**Pull request by hectorgon**
@deerzq @asaluja @Patrick1860

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala [hunk 1]
```diff
@@ -214,6 +207,9 @@ object FullRankLinearTrainer {
               val scores = model.scoreFlatFeature(flatFeatures)
               val posScore = scores.values(posIdx)
               val negScore = scores.values(negIdx)
+              // loss = max(0, margin + w(-) x - w(+) * pos)
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: loss = max(0, margin + w(-) * x - w(+) * x)

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala [hunk 2]
```diff
@@ -214,6 +207,9 @@ object FullRankLinearTrainer {
...
+              // so dloss / dw(-) = x and dloss / dw(+) = -x for hinge loss
+              // and dloss / dw(-) = regular hinge loss * x for squared hinge loss
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: => // and dloss / dw(-) = loss * x for squared hinge loss

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala [hunk 3]
```diff
@@ -174,14 +166,15 @@ object FullRankLinearTrainer {
       gradient.iterator
     })
     .reduceByKey((a, b) => sumGradients(a,b))
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: should we call GradientUtils.sumGradients(a,b) to be consistent and more clear about where the function come from ?

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala [hunk 4]
```diff
@@ -248,7 +245,8 @@ object FullRankLinearTrainer {
         gradient.iterator
       })
       .reduceByKey((a, b) => sumGradients(a,b))
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: same here

File: training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                               - lambda.toFloat * weight.values(i) // Regularization term.
+          }
+        }
+      }
+    })
+    val stepNorms = step.map(kv => math.sqrt(kv._2.dot(kv._2)))
+    log.info("Sum of Step L2 norms = " + stepNorms.reduce(_ + _))
+    val median = stepNorms.toBuffer.sorted.get(stepNorms.size / 2)
+    log.info("Median Step L2 norms = " + median)
+  }
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala]: nicely implemented.

**Comment by Patrick1860** [comment]
Have some minor style comments. The rprop implementation LGTM.

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala (hunk 1 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: Fixed

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala (hunk 2 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: Fixed

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala (hunk 3 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: Fixed

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala (hunk 4 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: Fixed

File: training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            tmp.setConstant(deltaInitial)
+            step.put(key, tmp)
+          }
+          val currStep = step.get(key).get
+          for (i <- 0 until dim) {
+            val prod = prevGrad.values(i) * gradient.grad.values(i)
+            if (prod > 0) {
+              currStep.values(i) = math.min(currStep.values(i) * etaPlus, deltaMax)
+            } else if (prod < 0) {
+              currStep.values(i) = math.max(currStep.values(i) * etaMinus, deltaMin)
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala]: I am not very familiar with this algorithm, but according to the paper, when `prod < 0`, weight should not be updated: `Additionally, in case of a change in sign, there should be no adaptation in the succeeding learning step`

Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala]: Ah I am using a newer variant called iRPROP- implemented here:
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.17.1332&rep=rep1&type=pdf

Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala]: Or rather algorithm 4 in this paper http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.17.1332&rep=rep1&type=pdf
